cirq-core/cirq/optimizers/two_qubit_to_sqrt_iswap.py

```diff
@@ -0,0 +1,413 @@
+            kak.single_qubit_operations_after[1] @ kak.single_qubit_operations_before[1],
+        )
+    ], kak.global_phase
+
+
+def _decomp_1sqrt_iswap_matrices(
+    kak: 'cirq.KakDecomposition',
+    atol: float = 1e-8,
+) -> Tuple[Sequence[Tuple[np.ndarray, np.ndarray]], complex]:
+    """Returns the single-qubit matrices for the 1-SQRT_ISWAP decomposition.
+
+    Assumes canonical x, y, z and (x, y, z) = (π/8, π/8, 0) within tolerance.
+    """
+    return [  # Pairs of single-qubit unitaries, SQRT_ISWAP between each is implied
+        kak.single_qubit_operations_before,
+        kak.single_qubit_operations_after,
+    ], kak.global_phase
+
+
+def _decomp_2sqrt_iswap_matrices(
+    kak: 'cirq.KakDecomposition',
+    atol: float = 1e-8,
+) -> Tuple[Sequence[Tuple[np.ndarray, np.ndarray]], complex]:
+    """Returns the single-qubit matrices for the 2-SQRT_ISWAP decomposition.
+
+    Assumes canonical x, y, z and x >= y + |z| within tolerance.  For x, y, z
+    that violate this inequality, three sqrt-iSWAP gates are required.
+
+    References:
+        Towards ultra-high fidelity quantum operations: SQiSW gate as a native
+        two-qubit gate
+        https://arxiv.org/abs/2105.06074
+    """
+    x, y, z = kak.interaction_coefficients
+    a0, a1 = kak.single_qubit_operations_before
+    c0, c1 = kak.single_qubit_operations_after
+
+    # Computed gate parameters: Eq. 4, 6, 7, 8 of the paper
+    # range limits added for robustness to numerical error
+    def safe_arccos(v):
+        return np.arccos(np.clip(v, -1, 1))
+
+    def nonzero_sign(v):
+        return -1 if v < 0 else 1
+
+    _c = np.clip(
+        np.sin(x + y - z) * np.sin(x - y + z) * np.sin(-x - y - z) * np.sin(-x + y + z), 0, 1
+    )
+    alpha = safe_arccos(np.cos(2 * x) - np.cos(2 * y) + np.cos(2 * z) + 2 * np.sqrt(_c))
+    beta = safe_arccos(np.cos(2 * x) - np.cos(2 * y) + np.cos(2 * z) - 2 * np.sqrt(_c))
+    _4ccs = 4 * (np.cos(x) * np.cos(z) * np.sin(y)) ** 2  # Intermediate value
+    gamma = safe_arccos(
+        nonzero_sign(z)
+        * np.sqrt(_4ccs / (_4ccs + np.clip(np.cos(2 * x) * np.cos(2 * y) * np.cos(2 * z), 0, 1)))
+    )
+
+    # Inner single-qubit gates: Fig. 4 of the paper
+    b0 = (
+        protocols.unitary(ops.rz(-gamma))
+        @ protocols.unitary(ops.rx(-alpha))
+        @ protocols.unitary(ops.rz(-gamma))
+    )
+    b1 = protocols.unitary(ops.rx(-beta))
+
+    # Compute KAK on the decomposition to determine outer single-qubit gates
+    # There is no known closed form solution for these gates
+    u_sqrt_iswap = protocols.unitary(ops.SQRT_ISWAP)
+    u = u_sqrt_iswap @ np.kron(b0, b1) @ u_sqrt_iswap  # Unitary of decomposition
+    kak_fix = linalg.kak_decomposition(u, atol=atol / 10, rtol=0, check_preconditions=False)
+    d0, d1 = kak_fix.single_qubit_operations_before
+    e0, e1 = kak_fix.single_qubit_operations_after
+
+    return [  # Pairs of single-qubit unitaries, SQRT_ISWAP between each is implied
```

Comment:
Do you mind keeping the notation from the paper? 
I.e. a, b for outer single qubit gates, c for the inner  and d,e for the kak fix.

![image](https://user-images.githubusercontent.com/635073/124219983-188c5b80-dacb-11eb-8bfa-f3b40245dfd8.png)

Reply:
I would also mention in a comment that this follows the first branch of `procedure DECOMP(U)` in `Algorithm 1`.

Reply:
Sure.